rosidl_parser/rosidl_parser/definition.py

```diff
@@ -107,170 +167,231 @@ def __eq__(self, other):
         return super().__eq__(other) and self.name == other.name
 
 
-class NamespacedType(BaseType):
+class NamespacedType(AbstractNestableType):
     """A type identified by a name in a namespaced scope."""
 
     __slots__ = ('namespaces', 'name')
 
-    def __init__(self, namespaces, name):
+    def __init__(self, namespaces: Iterable[str], name: str):
         """
         Constructor.
 
-        :param list[str] namespaces: the names of nested namespaces identifying
-          a specific scope
-        :param str name: the name of the type within that scope
+        :param namespaces: the names of nested namespaces identifying a
+          specific scope
+        :param name: the name of the type within that scope
         """
         super().__init__()
         self.namespaces = namespaces
         self.name = name
 
+    def namespaced_name(self) -> Tuple[str, ...]:
+        return (*self.namespaces, self.name)
+
     def __eq__(self, other):
         return super().__eq__(other) and \
             self.namespaces == other.namespaces and self.name == other.name
 
 
-class BaseString(BaseType):
-    """The base class of string types."""
+class AbstractGenericString(AbstractNestableType):
+    """The abstract base class of all string types."""
+
+    __slots__ = ()
+
+    def has_maximum_size(self):
+        raise NotImplementedError('Only implemented in subclasses')
```

Comment:
The type hierarchy can only provide one dimension to differentiate types. But in some cases the character type of a string (8-bit vs. 16-bit) isn't relevant but the logic wants to treat bounded and unbounded strings separately. While it could check for multiple types this method provides an easy way to distinguish this second dimension.

Reply:
While it saves a little bit of code the downside is that the calling code needs to explicitly import the function. In favor of keeping the calling code simpler I would prefer spending the extra lines in this code.